{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "RMDL_fmnist.ipynb",
      "provenance": [],
      "collapsed_sections": [],
      "machine_shape": "hm"
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "accelerator": "GPU"
  },
  "cells": [
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "g1g6iXwKJvwB",
        "outputId": "88c739ff-e52c-4cbc-c949-5df9544b5254"
      },
      "source": [
        "!pip install RMDL"
      ],
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Requirement already satisfied: RMDL in /usr/local/lib/python3.6/dist-packages (1.0.8)\n",
            "Requirement already satisfied: scikit-learn>=0.19.0 in /usr/local/lib/python3.6/dist-packages (from RMDL) (0.22.2.post1)\n",
            "Requirement already satisfied: nltk>=3.2.4 in /usr/local/lib/python3.6/dist-packages (from RMDL) (3.2.5)\n",
            "Requirement already satisfied: numpy>=1.12.1 in /usr/local/lib/python3.6/dist-packages (from RMDL) (1.19.5)\n",
            "Requirement already satisfied: pandas>=0.22.0 in /usr/local/lib/python3.6/dist-packages (from RMDL) (1.1.5)\n",
            "Requirement already satisfied: matplotlib>=2.1.2 in /usr/local/lib/python3.6/dist-packages (from RMDL) (3.2.2)\n",
            "Requirement already satisfied: scipy in /usr/local/lib/python3.6/dist-packages (from RMDL) (1.4.1)\n",
            "Requirement already satisfied: keras>=2.0.9 in /usr/local/lib/python3.6/dist-packages (from RMDL) (2.4.3)\n",
            "Requirement already satisfied: tensorflow in /usr/local/lib/python3.6/dist-packages (from RMDL) (2.4.1)\n",
            "Requirement already satisfied: joblib>=0.11 in /usr/local/lib/python3.6/dist-packages (from scikit-learn>=0.19.0->RMDL) (1.0.0)\n",
            "Requirement already satisfied: six in /usr/local/lib/python3.6/dist-packages (from nltk>=3.2.4->RMDL) (1.15.0)\n",
            "Requirement already satisfied: python-dateutil>=2.7.3 in /usr/local/lib/python3.6/dist-packages (from pandas>=0.22.0->RMDL) (2.8.1)\n",
            "Requirement already satisfied: pytz>=2017.2 in /usr/local/lib/python3.6/dist-packages (from pandas>=0.22.0->RMDL) (2018.9)\n",
            "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.6/dist-packages (from matplotlib>=2.1.2->RMDL) (0.10.0)\n",
            "Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.6/dist-packages (from matplotlib>=2.1.2->RMDL) (1.3.1)\n",
            "Requirement already satisfied: pyparsing!=2.0.4,!=2.1.2,!=2.1.6,>=2.0.1 in /usr/local/lib/python3.6/dist-packages (from matplotlib>=2.1.2->RMDL) (2.4.7)\n",
            "Requirement already satisfied: pyyaml in /usr/local/lib/python3.6/dist-packages (from keras>=2.0.9->RMDL) (3.13)\n",
            "Requirement already satisfied: h5py in /usr/local/lib/python3.6/dist-packages (from keras>=2.0.9->RMDL) (2.10.0)\n",
            "Requirement already satisfied: opt-einsum~=3.3.0 in /usr/local/lib/python3.6/dist-packages (from tensorflow->RMDL) (3.3.0)\n",
            "Requirement already satisfied: termcolor~=1.1.0 in /usr/local/lib/python3.6/dist-packages (from tensorflow->RMDL) (1.1.0)\n",
            "Requirement already satisfied: astunparse~=1.6.3 in /usr/local/lib/python3.6/dist-packages (from tensorflow->RMDL) (1.6.3)\n",
            "Requirement already satisfied: wheel~=0.35 in /usr/local/lib/python3.6/dist-packages (from tensorflow->RMDL) (0.36.2)\n",
            "Requirement already satisfied: tensorboard~=2.4 in /usr/local/lib/python3.6/dist-packages (from tensorflow->RMDL) (2.4.1)\n",
            "Requirement already satisfied: grpcio~=1.32.0 in /usr/local/lib/python3.6/dist-packages (from tensorflow->RMDL) (1.32.0)\n",
            "Requirement already satisfied: wrapt~=1.12.1 in /usr/local/lib/python3.6/dist-packages (from tensorflow->RMDL) (1.12.1)\n",
            "Requirement already satisfied: keras-preprocessing~=1.1.2 in /usr/local/lib/python3.6/dist-packages (from tensorflow->RMDL) (1.1.2)\n",
            "Requirement already satisfied: tensorflow-estimator<2.5.0,>=2.4.0 in /usr/local/lib/python3.6/dist-packages (from tensorflow->RMDL) (2.4.0)\n",
            "Requirement already satisfied: absl-py~=0.10 in /usr/local/lib/python3.6/dist-packages (from tensorflow->RMDL) (0.10.0)\n",
            "Requirement already satisfied: google-pasta~=0.2 in /usr/local/lib/python3.6/dist-packages (from tensorflow->RMDL) (0.2.0)\n",
            "Requirement already satisfied: typing-extensions~=3.7.4 in /usr/local/lib/python3.6/dist-packages (from tensorflow->RMDL) (3.7.4.3)\n",
            "Requirement already satisfied: flatbuffers~=1.12.0 in /usr/local/lib/python3.6/dist-packages (from tensorflow->RMDL) (1.12)\n",
            "Requirement already satisfied: protobuf>=3.9.2 in /usr/local/lib/python3.6/dist-packages (from tensorflow->RMDL) (3.12.4)\n",
            "Requirement already satisfied: gast==0.3.3 in /usr/local/lib/python3.6/dist-packages (from tensorflow->RMDL) (0.3.3)\n",
            "Requirement already satisfied: google-auth<2,>=1.6.3 in /usr/local/lib/python3.6/dist-packages (from tensorboard~=2.4->tensorflow->RMDL) (1.25.0)\n",
            "Requirement already satisfied: tensorboard-plugin-wit>=1.6.0 in /usr/local/lib/python3.6/dist-packages (from tensorboard~=2.4->tensorflow->RMDL) (1.8.0)\n",
            "Requirement already satisfied: setuptools>=41.0.0 in /usr/local/lib/python3.6/dist-packages (from tensorboard~=2.4->tensorflow->RMDL) (53.0.0)\n",
            "Requirement already satisfied: markdown>=2.6.8 in /usr/local/lib/python3.6/dist-packages (from tensorboard~=2.4->tensorflow->RMDL) (3.3.3)\n",
            "Requirement already satisfied: requests<3,>=2.21.0 in /usr/local/lib/python3.6/dist-packages (from tensorboard~=2.4->tensorflow->RMDL) (2.23.0)\n",
            "Requirement already satisfied: werkzeug>=0.11.15 in /usr/local/lib/python3.6/dist-packages (from tensorboard~=2.4->tensorflow->RMDL) (1.0.1)\n",
            "Requirement already satisfied: google-auth-oauthlib<0.5,>=0.4.1 in /usr/local/lib/python3.6/dist-packages (from tensorboard~=2.4->tensorflow->RMDL) (0.4.2)\n",
            "Requirement already satisfied: cachetools<5.0,>=2.0.0 in /usr/local/lib/python3.6/dist-packages (from google-auth<2,>=1.6.3->tensorboard~=2.4->tensorflow->RMDL) (4.2.1)\n",
            "Requirement already satisfied: rsa<5,>=3.1.4; python_version >= \"3.6\" in /usr/local/lib/python3.6/dist-packages (from google-auth<2,>=1.6.3->tensorboard~=2.4->tensorflow->RMDL) (4.7)\n",
            "Requirement already satisfied: pyasn1-modules>=0.2.1 in /usr/local/lib/python3.6/dist-packages (from google-auth<2,>=1.6.3->tensorboard~=2.4->tensorflow->RMDL) (0.2.8)\n",
            "Requirement already satisfied: importlib-metadata; python_version < \"3.8\" in /usr/local/lib/python3.6/dist-packages (from markdown>=2.6.8->tensorboard~=2.4->tensorflow->RMDL) (3.4.0)\n",
            "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /usr/local/lib/python3.6/dist-packages (from requests<3,>=2.21.0->tensorboard~=2.4->tensorflow->RMDL) (1.24.3)\n",
            "Requirement already satisfied: chardet<4,>=3.0.2 in /usr/local/lib/python3.6/dist-packages (from requests<3,>=2.21.0->tensorboard~=2.4->tensorflow->RMDL) (3.0.4)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.6/dist-packages (from requests<3,>=2.21.0->tensorboard~=2.4->tensorflow->RMDL) (2020.12.5)\n",
            "Requirement already satisfied: idna<3,>=2.5 in /usr/local/lib/python3.6/dist-packages (from requests<3,>=2.21.0->tensorboard~=2.4->tensorflow->RMDL) (2.10)\n",
            "Requirement already satisfied: requests-oauthlib>=0.7.0 in /usr/local/lib/python3.6/dist-packages (from google-auth-oauthlib<0.5,>=0.4.1->tensorboard~=2.4->tensorflow->RMDL) (1.3.0)\n",
            "Requirement already satisfied: pyasn1>=0.1.3 in /usr/local/lib/python3.6/dist-packages (from rsa<5,>=3.1.4; python_version >= \"3.6\"->google-auth<2,>=1.6.3->tensorboard~=2.4->tensorflow->RMDL) (0.4.8)\n",
            "Requirement already satisfied: zipp>=0.5 in /usr/local/lib/python3.6/dist-packages (from importlib-metadata; python_version < \"3.8\"->markdown>=2.6.8->tensorboard~=2.4->tensorflow->RMDL) (3.4.0)\n",
            "Requirement already satisfied: oauthlib>=3.0.0 in /usr/local/lib/python3.6/dist-packages (from requests-oauthlib>=0.7.0->google-auth-oauthlib<0.5,>=0.4.1->tensorboard~=2.4->tensorflow->RMDL) (3.1.0)\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "cRoUM6FEKXmK"
      },
      "source": [
        "from keras.datasets import fashion_mnist\n",
        "import numpy as np\n",
        "from RMDL import RMDL_Image as RMDL"
      ],
      "execution_count": 9,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "knOkzha2Klgm",
        "outputId": "46f09dbb-8177-4c06-8f47-1784424ee570"
      },
      "source": [
        "(X_train, y_train), (X_test, y_test) = fashion_mnist.load_data()\n",
        "X_train_D = X_train.reshape(X_train.shape[0], 28, 28, 1).astype('float32')\n",
        "X_test_D = X_test.reshape(X_test.shape[0], 28, 28, 1).astype('float32')\n",
        "X_train = X_train_D / 255.0\n",
        "X_test = X_test_D / 255.0\n",
        "number_of_classes = np.unique(y_train).shape[0]\n",
        "shape = (28, 28, 1)"
      ],
      "execution_count": 10,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Downloading data from https://storage.googleapis.com/tensorflow/tf-keras-datasets/train-labels-idx1-ubyte.gz\n",
            "32768/29515 [=================================] - 0s 0us/step\n",
            "Downloading data from https://storage.googleapis.com/tensorflow/tf-keras-datasets/train-images-idx3-ubyte.gz\n",
            "26427392/26421880 [==============================] - 0s 0us/step\n",
            "Downloading data from https://storage.googleapis.com/tensorflow/tf-keras-datasets/t10k-labels-idx1-ubyte.gz\n",
            "8192/5148 [===============================================] - 0s 0us/step\n",
            "Downloading data from https://storage.googleapis.com/tensorflow/tf-keras-datasets/t10k-images-idx3-ubyte.gz\n",
            "4423680/4422102 [==============================] - 0s 0us/step\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "5BkMHRa-KnV4",
        "outputId": "c17882a7-dc1e-4818-f905-8dc551aed768"
      },
      "source": [
        "batch_size = 128\n",
        "sparse_categorical = 0\n",
        "n_epochs = [10, 10, 10]  ## DNN-RNN-CNN\n",
        "Random_Deep = [2, 2, 2]  ## DNN-RNN-CNN\n",
        "\n",
        "RMDL.Image_Classification(X_train, y_train, X_test, y_test,shape,\n",
        "                     batch_size=batch_size,\n",
        "                     sparse_categorical=True,\n",
        "                     random_deep=Random_Deep,\n",
        "                     epochs=n_epochs)"
      ],
      "execution_count": 11,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "DNN  0 \n",
            "\n",
            "(28, 28, 1)\n",
            "<tensorflow.python.keras.optimizer_v2.adagrad.Adagrad object at 0x7fb600195e10>\n",
            "Epoch 1/10\n",
            "469/469 - 3s - loss: 1.4116 - accuracy: 0.5996 - val_loss: 1.0340 - val_accuracy: 0.6962\n",
            "\n",
            "Epoch 00001: val_accuracy improved from -inf to 0.69620, saving model to weights\\weights_DNN_0.hdf5\n",
            "Epoch 2/10\n",
            "469/469 - 1s - loss: 0.9174 - accuracy: 0.7232 - val_loss: 0.8440 - val_accuracy: 0.7347\n",
            "\n",
            "Epoch 00002: val_accuracy improved from 0.69620 to 0.73470, saving model to weights\\weights_DNN_0.hdf5\n",
            "Epoch 3/10\n",
            "469/469 - 1s - loss: 0.7923 - accuracy: 0.7534 - val_loss: 0.7611 - val_accuracy: 0.7588\n",
            "\n",
            "Epoch 00003: val_accuracy improved from 0.73470 to 0.75880, saving model to weights\\weights_DNN_0.hdf5\n",
            "Epoch 4/10\n",
            "469/469 - 1s - loss: 0.7261 - accuracy: 0.7707 - val_loss: 0.7101 - val_accuracy: 0.7715\n",
            "\n",
            "Epoch 00004: val_accuracy improved from 0.75880 to 0.77150, saving model to weights\\weights_DNN_0.hdf5\n",
            "Epoch 5/10\n",
            "469/469 - 1s - loss: 0.6835 - accuracy: 0.7830 - val_loss: 0.6754 - val_accuracy: 0.7811\n",
            "\n",
            "Epoch 00005: val_accuracy improved from 0.77150 to 0.78110, saving model to weights\\weights_DNN_0.hdf5\n",
            "Epoch 6/10\n",
            "469/469 - 1s - loss: 0.6518 - accuracy: 0.7929 - val_loss: 0.6490 - val_accuracy: 0.7894\n",
            "\n",
            "Epoch 00006: val_accuracy improved from 0.78110 to 0.78940, saving model to weights\\weights_DNN_0.hdf5\n",
            "Epoch 7/10\n",
            "469/469 - 1s - loss: 0.6287 - accuracy: 0.7983 - val_loss: 0.6287 - val_accuracy: 0.7949\n",
            "\n",
            "Epoch 00007: val_accuracy improved from 0.78940 to 0.79490, saving model to weights\\weights_DNN_0.hdf5\n",
            "Epoch 8/10\n",
            "469/469 - 1s - loss: 0.6081 - accuracy: 0.8039 - val_loss: 0.6117 - val_accuracy: 0.7991\n",
            "\n",
            "Epoch 00008: val_accuracy improved from 0.79490 to 0.79910, saving model to weights\\weights_DNN_0.hdf5\n",
            "Epoch 9/10\n",
            "469/469 - 1s - loss: 0.5934 - accuracy: 0.8091 - val_loss: 0.5973 - val_accuracy: 0.8025\n",
            "\n",
            "Epoch 00009: val_accuracy improved from 0.79910 to 0.80250, saving model to weights\\weights_DNN_0.hdf5\n",
            "Epoch 10/10\n",
            "469/469 - 1s - loss: 0.5789 - accuracy: 0.8128 - val_loss: 0.5858 - val_accuracy: 0.8073\n",
            "\n",
            "Epoch 00010: val_accuracy improved from 0.80250 to 0.80730, saving model to weights\\weights_DNN_0.hdf5\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "stream",
          "text": [
            "/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/engine/sequential.py:450: UserWarning: `model.predict_classes()` is deprecated and will be removed after 2021-01-01. Please use instead:* `np.argmax(model.predict(x), axis=-1)`,   if your model does multi-class classification   (e.g. if it uses a `softmax` last-layer activation).* `(model.predict(x) > 0.5).astype(\"int32\")`,   if your model does binary classification   (e.g. if it uses a `sigmoid` last-layer activation).\n",
            "  warnings.warn('`model.predict_classes()` is deprecated and '\n"
          ],
          "name": "stderr"
        },
        {
          "output_type": "stream",
          "text": [
            "DNN  1 \n",
            "\n",
            "(28, 28, 1)\n",
            "<tensorflow.python.keras.optimizer_v2.rmsprop.RMSprop object at 0x7fb59b6d1ba8>\n",
            "Epoch 1/10\n",
            "469/469 - 1s - loss: 0.5635 - accuracy: 0.8007 - val_loss: 0.5334 - val_accuracy: 0.8131\n",
            "\n",
            "Epoch 00001: val_accuracy improved from -inf to 0.81310, saving model to weights\\weights_DNN_1.hdf5\n",
            "Epoch 2/10\n",
            "469/469 - 1s - loss: 0.3923 - accuracy: 0.8567 - val_loss: 0.4099 - val_accuracy: 0.8506\n",
            "\n",
            "Epoch 00002: val_accuracy improved from 0.81310 to 0.85060, saving model to weights\\weights_DNN_1.hdf5\n",
            "Epoch 3/10\n",
            "469/469 - 1s - loss: 0.3466 - accuracy: 0.8741 - val_loss: 0.4019 - val_accuracy: 0.8507\n",
            "\n",
            "Epoch 00003: val_accuracy improved from 0.85060 to 0.85070, saving model to weights\\weights_DNN_1.hdf5\n",
            "Epoch 4/10\n",
            "469/469 - 1s - loss: 0.3211 - accuracy: 0.8840 - val_loss: 0.3802 - val_accuracy: 0.8593\n",
            "\n",
            "Epoch 00004: val_accuracy improved from 0.85070 to 0.85930, saving model to weights\\weights_DNN_1.hdf5\n",
            "Epoch 5/10\n",
            "469/469 - 1s - loss: 0.3027 - accuracy: 0.8895 - val_loss: 0.3741 - val_accuracy: 0.8644\n",
            "\n",
            "Epoch 00005: val_accuracy improved from 0.85930 to 0.86440, saving model to weights\\weights_DNN_1.hdf5\n",
            "Epoch 6/10\n",
            "469/469 - 1s - loss: 0.2897 - accuracy: 0.8943 - val_loss: 0.3531 - val_accuracy: 0.8753\n",
            "\n",
            "Epoch 00006: val_accuracy improved from 0.86440 to 0.87530, saving model to weights\\weights_DNN_1.hdf5\n",
            "Epoch 7/10\n",
            "469/469 - 1s - loss: 0.2763 - accuracy: 0.8985 - val_loss: 0.3393 - val_accuracy: 0.8821\n",
            "\n",
            "Epoch 00007: val_accuracy improved from 0.87530 to 0.88210, saving model to weights\\weights_DNN_1.hdf5\n",
            "Epoch 8/10\n",
            "469/469 - 1s - loss: 0.2645 - accuracy: 0.9021 - val_loss: 0.3467 - val_accuracy: 0.8784\n",
            "\n",
            "Epoch 00008: val_accuracy did not improve from 0.88210\n",
            "Epoch 9/10\n",
            "469/469 - 1s - loss: 0.2567 - accuracy: 0.9054 - val_loss: 0.3682 - val_accuracy: 0.8738\n",
            "\n",
            "Epoch 00009: val_accuracy did not improve from 0.88210\n",
            "Epoch 10/10\n",
            "469/469 - 1s - loss: 0.2466 - accuracy: 0.9081 - val_loss: 0.3383 - val_accuracy: 0.8825\n",
            "\n",
            "Epoch 00010: val_accuracy improved from 0.88210 to 0.88250, saving model to weights\\weights_DNN_1.hdf5\n",
            "RNN  0 \n",
            "\n",
            "WARNING:tensorflow:Layer lstm will not use cuDNN kernel since it doesn't meet the cuDNN kernel criteria. It will use generic GPU kernel as fallback when running on GPU\n",
            "WARNING:tensorflow:Layer lstm_1 will not use cuDNN kernel since it doesn't meet the cuDNN kernel criteria. It will use generic GPU kernel as fallback when running on GPU\n",
            "<tensorflow.python.keras.optimizer_v2.rmsprop.RMSprop object at 0x7fb59a47e7b8>\n",
            "Epoch 1/10\n",
            "469/469 - 69s - loss: 1.0069 - accuracy: 0.6225 - val_loss: 0.6700 - val_accuracy: 0.7567\n",
            "\n",
            "Epoch 00001: val_accuracy improved from -inf to 0.75670, saving model to weights\\weights_RNN_0.hdf5\n",
            "Epoch 2/10\n",
            "469/469 - 67s - loss: 0.6336 - accuracy: 0.7665 - val_loss: 0.5531 - val_accuracy: 0.7879\n",
            "\n",
            "Epoch 00002: val_accuracy improved from 0.75670 to 0.78790, saving model to weights\\weights_RNN_0.hdf5\n",
            "Epoch 3/10\n",
            "469/469 - 66s - loss: 0.5192 - accuracy: 0.8076 - val_loss: 0.4714 - val_accuracy: 0.8254\n",
            "\n",
            "Epoch 00003: val_accuracy improved from 0.78790 to 0.82540, saving model to weights\\weights_RNN_0.hdf5\n",
            "Epoch 4/10\n",
            "469/469 - 65s - loss: 0.4533 - accuracy: 0.8325 - val_loss: 0.4642 - val_accuracy: 0.8266\n",
            "\n",
            "Epoch 00004: val_accuracy improved from 0.82540 to 0.82660, saving model to weights\\weights_RNN_0.hdf5\n",
            "Epoch 5/10\n",
            "469/469 - 66s - loss: 0.4165 - accuracy: 0.8451 - val_loss: 0.4177 - val_accuracy: 0.8442\n",
            "\n",
            "Epoch 00005: val_accuracy improved from 0.82660 to 0.84420, saving model to weights\\weights_RNN_0.hdf5\n",
            "Epoch 6/10\n",
            "469/469 - 65s - loss: 0.3886 - accuracy: 0.8554 - val_loss: 0.4049 - val_accuracy: 0.8526\n",
            "\n",
            "Epoch 00006: val_accuracy improved from 0.84420 to 0.85260, saving model to weights\\weights_RNN_0.hdf5\n",
            "Epoch 7/10\n",
            "469/469 - 66s - loss: 0.3663 - accuracy: 0.8636 - val_loss: 0.3881 - val_accuracy: 0.8557\n",
            "\n",
            "Epoch 00007: val_accuracy improved from 0.85260 to 0.85570, saving model to weights\\weights_RNN_0.hdf5\n",
            "Epoch 8/10\n",
            "469/469 - 67s - loss: 0.3510 - accuracy: 0.8681 - val_loss: 0.3866 - val_accuracy: 0.8590\n",
            "\n",
            "Epoch 00008: val_accuracy improved from 0.85570 to 0.85900, saving model to weights\\weights_RNN_0.hdf5\n",
            "Epoch 9/10\n",
            "469/469 - 66s - loss: 0.3348 - accuracy: 0.8752 - val_loss: 0.3506 - val_accuracy: 0.8688\n",
            "\n",
            "Epoch 00009: val_accuracy improved from 0.85900 to 0.86880, saving model to weights\\weights_RNN_0.hdf5\n",
            "Epoch 10/10\n",
            "469/469 - 66s - loss: 0.3241 - accuracy: 0.8781 - val_loss: 0.3735 - val_accuracy: 0.8583\n",
            "\n",
            "Epoch 00010: val_accuracy did not improve from 0.86880\n",
            "RNN  1 \n",
            "\n",
            "WARNING:tensorflow:Layer lstm_2 will not use cuDNN kernel since it doesn't meet the cuDNN kernel criteria. It will use generic GPU kernel as fallback when running on GPU\n",
            "WARNING:tensorflow:Layer lstm_3 will not use cuDNN kernel since it doesn't meet the cuDNN kernel criteria. It will use generic GPU kernel as fallback when running on GPU\n",
            "<tensorflow.python.keras.optimizer_v2.adagrad.Adagrad object at 0x7fb59294f780>\n",
            "Epoch 1/10\n",
            "469/469 - 70s - loss: 2.3019 - accuracy: 0.1226 - val_loss: 2.3006 - val_accuracy: 0.1925\n",
            "\n",
            "Epoch 00001: val_accuracy improved from -inf to 0.19250, saving model to weights\\weights_RNN_1.hdf5\n",
            "Epoch 2/10\n",
            "469/469 - 67s - loss: 2.2995 - accuracy: 0.1965 - val_loss: 2.2980 - val_accuracy: 0.2144\n",
            "\n",
            "Epoch 00002: val_accuracy improved from 0.19250 to 0.21440, saving model to weights\\weights_RNN_1.hdf5\n",
            "Epoch 3/10\n",
            "469/469 - 65s - loss: 2.2967 - accuracy: 0.2009 - val_loss: 2.2950 - val_accuracy: 0.1467\n",
            "\n",
            "Epoch 00003: val_accuracy did not improve from 0.21440\n",
            "Epoch 4/10\n",
            "469/469 - 66s - loss: 2.2932 - accuracy: 0.1392 - val_loss: 2.2908 - val_accuracy: 0.1524\n",
            "\n",
            "Epoch 00004: val_accuracy did not improve from 0.21440\n",
            "Epoch 5/10\n",
            "469/469 - 64s - loss: 2.2882 - accuracy: 0.1561 - val_loss: 2.2845 - val_accuracy: 0.1503\n",
            "\n",
            "Epoch 00005: val_accuracy did not improve from 0.21440\n",
            "Epoch 6/10\n",
            "469/469 - 65s - loss: 2.2801 - accuracy: 0.1791 - val_loss: 2.2739 - val_accuracy: 0.1911\n",
            "\n",
            "Epoch 00006: val_accuracy did not improve from 0.21440\n",
            "Epoch 7/10\n",
            "469/469 - 64s - loss: 2.2652 - accuracy: 0.1960 - val_loss: 2.2527 - val_accuracy: 0.2017\n",
            "\n",
            "Epoch 00007: val_accuracy did not improve from 0.21440\n",
            "Epoch 8/10\n",
            "469/469 - 66s - loss: 2.2308 - accuracy: 0.1995 - val_loss: 2.1985 - val_accuracy: 0.1983\n",
            "\n",
            "Epoch 00008: val_accuracy did not improve from 0.21440\n",
            "Epoch 9/10\n",
            "469/469 - 66s - loss: 2.1494 - accuracy: 0.1971 - val_loss: 2.0989 - val_accuracy: 0.2092\n",
            "\n",
            "Epoch 00009: val_accuracy did not improve from 0.21440\n",
            "Epoch 10/10\n",
            "469/469 - 66s - loss: 2.0633 - accuracy: 0.2278 - val_loss: 2.0293 - val_accuracy: 0.2533\n",
            "\n",
            "Epoch 00010: val_accuracy improved from 0.21440 to 0.25330, saving model to weights\\weights_RNN_1.hdf5\n",
            "CNN  0 \n",
            "\n",
            "<tensorflow.python.keras.optimizer_v2.rmsprop.RMSprop object at 0x7fb5925762e8>\n",
            "Epoch 1/10\n",
            "469/469 - 23s - loss: 0.5647 - accuracy: 0.7995 - val_loss: 0.3416 - val_accuracy: 0.8743\n",
            "\n",
            "Epoch 00001: val_accuracy improved from -inf to 0.87430, saving model to weights\\weights_CNN_0.hdf5\n",
            "Epoch 2/10\n",
            "469/469 - 19s - loss: 0.2603 - accuracy: 0.9036 - val_loss: 0.2746 - val_accuracy: 0.8961\n",
            "\n",
            "Epoch 00002: val_accuracy improved from 0.87430 to 0.89610, saving model to weights\\weights_CNN_0.hdf5\n",
            "Epoch 3/10\n",
            "469/469 - 19s - loss: 0.2098 - accuracy: 0.9226 - val_loss: 0.2179 - val_accuracy: 0.9195\n",
            "\n",
            "Epoch 00003: val_accuracy improved from 0.89610 to 0.91950, saving model to weights\\weights_CNN_0.hdf5\n",
            "Epoch 4/10\n",
            "469/469 - 19s - loss: 0.1795 - accuracy: 0.9356 - val_loss: 0.2147 - val_accuracy: 0.9219\n",
            "\n",
            "Epoch 00004: val_accuracy improved from 0.91950 to 0.92190, saving model to weights\\weights_CNN_0.hdf5\n",
            "Epoch 5/10\n",
            "469/469 - 19s - loss: 0.1542 - accuracy: 0.9423 - val_loss: 0.2303 - val_accuracy: 0.9213\n",
            "\n",
            "Epoch 00005: val_accuracy did not improve from 0.92190\n",
            "Epoch 6/10\n",
            "469/469 - 19s - loss: 0.1332 - accuracy: 0.9518 - val_loss: 0.2273 - val_accuracy: 0.9225\n",
            "\n",
            "Epoch 00006: val_accuracy improved from 0.92190 to 0.92250, saving model to weights\\weights_CNN_0.hdf5\n",
            "Epoch 7/10\n",
            "469/469 - 19s - loss: 0.1175 - accuracy: 0.9572 - val_loss: 0.2371 - val_accuracy: 0.9240\n",
            "\n",
            "Epoch 00007: val_accuracy improved from 0.92250 to 0.92400, saving model to weights\\weights_CNN_0.hdf5\n",
            "Epoch 8/10\n",
            "469/469 - 19s - loss: 0.1071 - accuracy: 0.9617 - val_loss: 0.2273 - val_accuracy: 0.9260\n",
            "\n",
            "Epoch 00008: val_accuracy improved from 0.92400 to 0.92600, saving model to weights\\weights_CNN_0.hdf5\n",
            "Epoch 9/10\n",
            "469/469 - 19s - loss: 0.0978 - accuracy: 0.9657 - val_loss: 0.2407 - val_accuracy: 0.9305\n",
            "\n",
            "Epoch 00009: val_accuracy improved from 0.92600 to 0.93050, saving model to weights\\weights_CNN_0.hdf5\n",
            "Epoch 10/10\n",
            "469/469 - 19s - loss: 0.0886 - accuracy: 0.9681 - val_loss: 0.3180 - val_accuracy: 0.9270\n",
            "\n",
            "Epoch 00010: val_accuracy did not improve from 0.93050\n",
            "CNN  1 \n",
            "\n",
            "Error in model 1  try to re-generate another model\n",
            "CNN  1 \n",
            "\n",
            "Error in model 1  try to re-generate another model\n",
            "CNN  1 \n",
            "\n",
            "<tensorflow.python.keras.optimizer_v2.rmsprop.RMSprop object at 0x7fb571758ba8>\n",
            "Epoch 1/10\n",
            "469/469 - 15s - loss: 0.6069 - accuracy: 0.7757 - val_loss: 0.3888 - val_accuracy: 0.8553\n",
            "\n",
            "Epoch 00001: val_accuracy improved from -inf to 0.85530, saving model to weights\\weights_CNN_1.hdf5\n",
            "Epoch 2/10\n",
            "469/469 - 14s - loss: 0.2727 - accuracy: 0.8998 - val_loss: 0.2497 - val_accuracy: 0.9094\n",
            "\n",
            "Epoch 00002: val_accuracy improved from 0.85530 to 0.90940, saving model to weights\\weights_CNN_1.hdf5\n",
            "Epoch 3/10\n",
            "469/469 - 14s - loss: 0.2183 - accuracy: 0.9201 - val_loss: 0.2426 - val_accuracy: 0.9127\n",
            "\n",
            "Epoch 00003: val_accuracy improved from 0.90940 to 0.91270, saving model to weights\\weights_CNN_1.hdf5\n",
            "Epoch 4/10\n",
            "469/469 - 14s - loss: 0.1893 - accuracy: 0.9306 - val_loss: 0.2195 - val_accuracy: 0.9234\n",
            "\n",
            "Epoch 00004: val_accuracy improved from 0.91270 to 0.92340, saving model to weights\\weights_CNN_1.hdf5\n",
            "Epoch 5/10\n",
            "469/469 - 14s - loss: 0.1661 - accuracy: 0.9401 - val_loss: 0.2470 - val_accuracy: 0.9208\n",
            "\n",
            "Epoch 00005: val_accuracy did not improve from 0.92340\n",
            "Epoch 6/10\n",
            "469/469 - 14s - loss: 0.1507 - accuracy: 0.9459 - val_loss: 0.2773 - val_accuracy: 0.9234\n",
            "\n",
            "Epoch 00006: val_accuracy did not improve from 0.92340\n",
            "Epoch 7/10\n",
            "469/469 - 14s - loss: 0.1373 - accuracy: 0.9516 - val_loss: 0.2442 - val_accuracy: 0.9113\n",
            "\n",
            "Epoch 00007: val_accuracy did not improve from 0.92340\n",
            "Epoch 8/10\n",
            "469/469 - 14s - loss: 0.1297 - accuracy: 0.9545 - val_loss: 0.2494 - val_accuracy: 0.9222\n",
            "\n",
            "Epoch 00008: val_accuracy did not improve from 0.92340\n",
            "Epoch 9/10\n",
            "469/469 - 14s - loss: 0.1260 - accuracy: 0.9562 - val_loss: 0.2390 - val_accuracy: 0.9256\n",
            "\n",
            "Epoch 00009: val_accuracy improved from 0.92340 to 0.92560, saving model to weights\\weights_CNN_1.hdf5\n",
            "Epoch 10/10\n",
            "469/469 - 14s - loss: 0.1181 - accuracy: 0.9583 - val_loss: 0.2158 - val_accuracy: 0.9321\n",
            "\n",
            "Epoch 00010: val_accuracy improved from 0.92560 to 0.93210, saving model to weights\\weights_CNN_1.hdf5\n",
            "(10000, 6)\n",
            "(10000, 6)\n",
            "Accuracy of 6 models: [0.8073, 0.8825, 0.8688, 0.2533, 0.9305, 0.9321]\n",
            "Accuracy: 0.9084\n",
            "F1_Micro: 0.9084\n",
            "F1_Macro: 0.9082553479496551\n",
            "F1_weighted: 0.9082553479496551\n"
          ],
          "name": "stdout"
        }
      ]
    }
  ]
}